lab2+scanner/parser/LR.py

```diff
@@ -36,15 +44,120 @@ def canonical_collection(self):
         s0 = self.closure([('T', '.' + self.g.S)])
         self.g.enrich_grammar()
         set_of_states.append(s0)
-        for s in set_of_states:
+        set_of_gotos = []
+        for idx, s in enumerate(set_of_states):
             for p in s:
                 for i in range(len(p[1])):
                     if p[1][i] == '.' and i != len(p[1]) - 1:
-                        new_s = self.closure(self.goto(s, p[1][i+1]))
+                        if (p[1][i + 1] not in self.g.E) and (p[1][i + 1] not in self.g.N) and (
+                                p[1][i + 1] != self.g.S):
+                            sym = p[1][i + 1]
+                            j = i + 1
+                            while sym not in self.g.E and sym not in self.g.N and sym != self.g.S:
+                                sym += p[1][j + 1]
+                                j += 1
+                        else:
+                            sym = p[1][i + 1]
+                        new_s = self.closure(self.goto(s, sym, idx))
                         if new_s not in set_of_states:
+                            set_of_gotos.append((idx, sym, len(set_of_states)))
                             set_of_states.append(new_s)
-        return set_of_states
+                        else:
+                            set_of_gotos.append((idx, sym, set_of_states.index(new_s)))
+                        break
+        return set_of_states, set_of_gotos
 
     def print_canonical_collection(self, can_col):
         for i in range(len(can_col)):
             print("s" + str(i) + ":" + str(can_col[i]))
+
+    def compute_parsing_table(self, can_col):
+        parsing_table = {}
+        list_of_symbols = self.g.E + self.g.N
+        set_of_states = can_col[0]
+        set_of_gotos = can_col[1]
+        for idx_of_s, s in enumerate(set_of_states):
+            row_for_s = {'action': ''}
+            for symbol in list_of_symbols:
+                row_for_s[symbol] = 'error'
+            for elem in s:
+                if elem == ('T', 'S.') or elem == ('T', 'P.'):
```

Comment:
I think you should use the starting symbol from the grammar, it might not always be S or P .